man/clean_content_views.8
.\" Manpage for sat_import.
.\" Contact ggatward@redhat.com to correct errors or typos.
.TH CLEAN_CONTENT_VIEWS 8 "04 Jan 2017" "sat6_scripts" "sat6_scripts User Manual" man page"
.SH NAME
clean_content_views \- Remove unused content view versoions from Satellite 6

.SH SYNOPSIS
.B clean_content_views [\-o ORGANISATION] [\-k KEEP] [\-a] [\-d]
.LP
.B "clean_content_views --help"

.SH DESCRIPTION
.B clean_content_views
is a custom python script for cleaning unused versions of content views from Satellite 6.
Unused view versions older than the last promoted version are candidates for deletion, with the option to retain a given number of additional views.

.SH OPTIONS
The options that apply to the
.B clean_content_views
command are:
.PP
.BR "-o", " --org"
.I "ORGANISATION"
.RS 3
The organisation within Satellite to perform the activity for. The default organisation is defined in the
.I config.yml
file, but can be overridden with this option.
.RE
.PP
.BR "-a", " --all"
.RS 3
Clean all configured content views. By default
.B clean_content_views
will cleanup only the content views defined in the
.I config.yml
file - see
.BR sat6_scripts (8).
.RE
.PP
.BR "-i", " --ignorefirstpromoted"
.RS 3
Clean configured content views, ignoring the first promoted content view.
.PP
For example. a given content view contains versions 1 (goldenimg), 2, 3 (Production), 4, 5, 6 (Dev) and 7 (Library)
.br
In this example, versions 2, 4 and 5 are unused. Using the
.I -i
option (with a keep value of 1), versions 2 and 4 will be deleted and 5 will remain as a backup.
.RE
.PP
.BR "-k", " --keep"
.RS 3
Define the number of versions to keep beyond the oldest promoted version for all content views.
.PP
For example, a given content view contains versions 1, 2, 3 (Production), 4 (Quality), 5 and 6 (Library)
.br
In this example, versions 1, 2 and 5 are unused. By default, this script would delete versions 1 and 2 but not 5.
If the KEEP option is used to specify additional versions to keep (for example keep=1) then version 2 would not be deleted, as it is 1 greater than the oldest promoted version. This allows for a version rollback if required.
.PP
The number of old views to keep can also be defined per content view in the
.I config.yml
file - see
.BR sat6_scripts (8).
If the keep value is not specified in either the config.yml or the --keep option, the default value is 0.
.RE
.PP
.BR "-c", " --cleanall"
.RS 3
Clean all orphaned versions of the content view.
By default the removal of versions between published, in-use versions is not performed.
.PP
For example, a given content view contains versions 1, 2 (Production), 3, 4, 5, 6 (Quality), 7, 8 (Library)
.br
By default versions 3-5 and 7 will not be deleted as they exist in between 'live' published versions.
Using the
.B -c
option will cause these versions to be cleaned up as well as any older than the oldest promoted version.
.RE
.PP
.BR "-d", " --dryrun"
.RS 3
Do not perform the actual cleanup activity, but show what would be done.
This option can be used to verify that the correct configuration options have been defined and applied before running the actual cleanup of content.
.RE


.SH EXAMPLES
Clean all content views defined in the Satellite, globally leaving 1 older version:
.RS 6
.B "clean_content_views -a -k 1"
.RE

Show what will be cleaned by default, but do not actually remove the content:
.RS 6
.B "clean_content_views -d"
.RE

.SH FILES
.B Main Configuration
.RS 3
.I /usr/share/sat6_scripts/config/config.yml
.RE
.LP
.B Log Files
.RS 3
.I /var/log/satellite/*
.RE

.SH SEE ALSO
.BR sat6_scripts_config (8),
.BR publish_content_views (8),
.BR promote_content_views (8).

.SH AUTHOR
Geoff Gatward <ggatward@redhat.com>


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.
The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
- Option to tag a published content view with a custom comment
- Option to clean versions taking into account published versions

### Fixed
- Fix unreferenced var in check_disk_space function
- Various minor README corrections
- auto_content did not publish CV if the satellite is internet connected


## [1.2.1] - 2017-12-11
### Fixed
- Incorrect mapping of days in auto_content.py
- Support for username:token authentication to Artifactory Puppet Forge server
- Email now supports multiple recipients


## [1.2.0] - 2017-12-10
### Added
- push_puppetforge now supports jFrog Artifiactory repository via HTTP POST
- sat_import now checks for exports that have not been imported (missed/skipped)
- sat_import --fixhistory option to force align import/export histories
- Email notification capability for use when automating content scripts
- Add unattended option to allow scripts to be automated
- auto_content scripts to allow unattended import/publish/promote/clean activity

### Changed
- --notar export saved in /cdn_export dir rather than /export to prevent it being deleted

### Removed
- Skip GPG short option (-n)


## [1.1.1] - 2017-10-25
### Added
- Allow limiting of Publish and Promotion to specified number of CV's at once.


## [1.1.0] - 2017-10-23
### Added
- Proxy support for download_manifest
- Quiet option in publish/promote scripts
- Full repo package comparison written to log during disconnected import
- Provide a history of successful content exports and imports
- Present a warning if import has already been performed for given dataset
- Option to leave export in unarchived format (--notar)
- Option to force a disconnected (import only) Satellite to export if required

### Changed
- Updated exit codes throughout the scripts
- Updated export dataset name format to allow for multiple exports per day
- Updated to Semantic Versioning

### Fixed
- Various minor bugfixes and typos
- Fixed bug in DoV export

### Deprecated
- Skip GPG short option (-n) will be removed in 1.2.0. Long option (--nogpg) will remain.


## [1.0] - 2017-03-06
- Initial 'public' release

## 0.6 - 2017-02-27
- Last of a series of pre-release betas

[Unreleased]: https://github.com/RedHatSatellite/sat6_scripts/compare/1.2.1...HEAD
[1.2.1]: https://github.com/RedHatSatellite/sat6_scripts/compare/1.2.0...1.2.1
[1.2.0]: https://github.com/RedHatSatellite/sat6_scripts/compare/1.1.1...1.2.0
[1.1.1]: https://github.com/RedHatSatellite/sat6_scripts/compare/1.1.0...1.1.1
[1.1.0]: https://github.com/RedHatSatellite/sat6_scripts/compare/1.0...1.1.0
[1.0]: https://github.com/RedHatSatellite/sat6_scripts/compare/0.6...1.0


clean_content_views.py
#!/usr/bin/python
#title           :clean_content_views.py
#description     :Removes orphaned versions of Satellite 6 content views
#URL             :https://github.com/ggatward/sat6_scripts
#author          :Geoff Gatward <ggatward@redhat.com>
#notes           :This script is NOT SUPPORTED by Red Hat Global Support Services.
#license         :GPLv3
#==============================================================================
"""
Removes content view versions that don't belong to any environment

"""
#pylint: disable-msg=R0912,R0913,R0914,R0915

import sys, argparse
import simplejson as json
import helpers

def get_cv(org_id, cleanup_list, keep):
    """Get the content views"""

    # Query API to get all content views for our org
    cvs = helpers.get_json(
        helpers.KATELLO_API + "organizations/" + str(org_id) + "/content_views/")
    ver_list = {}
    ver_descr = {}
    ver_keep = {}

    for cv_result in cvs['results']:
        # We will never clean the DOV
        if cv_result['name'] != "Default Organization View":
            # Handle specific includes
            if cleanup_list:
                # The list contains dictionaries as elements. Process each dictionary
                for cv in cleanup_list:
                    # If the CV name does not appear in our config list, skip
                    if cv['view'] != cv_result['name']:
                        msg = "Skipping " + cv_result['name']
                        helpers.log_msg(msg, 'DEBUG')
                        continue
                    else:
                        msg = "Processing content view '" + cv_result['name'] + "' " \
                            + str(cv_result['id'])
                        helpers.log_msg(msg, 'DEBUG')

                        # Add the next version of the view, and how many versions to keep
                        ver_list[cv_result['id']] = cv_result['id']
                        ver_descr[cv_result['id']] = cv_result['name']
                        ver_keep[cv_result['id']] = cv['keep']

            # Handle the 'all' option
            else:
                msg = "Processing content view '" + cv_result['name'] + "' " \
                    + str(cv_result['id'])
                helpers.log_msg(msg, 'DEBUG')

                # Add the next version of the view, and how many versions to keep
                ver_list[cv_result['id']] = cv_result['id']
                ver_descr[cv_result['id']] = cv_result['name']
                ver_keep[cv_result['id']] = keep


    return ver_list, ver_descr, ver_keep


def get_content_view_info(cvid):
    """
    Return Content View Info for a given CV ID
    """
    cvinfo = helpers.get_json(
        helpers.KATELLO_API + "content_views/" + str(cvid))

    return cvinfo


def cleanup(ver_list, ver_descr, dry_run, runuser, ver_keep, cleanall, ignorefirstpromoted):
    """Clean Content Views"""

    # Set the task name to be displayed in the task monitoring stage
    task_name = "Cleanup content views"

    # Now we have all the info needed, we can actually trigger the cleanup.
    task_list = []
    ref_list = {}

    # Catch scenario that no CV versions are found matching cleanup criteria
    if not ver_list:
        msg = "No content view versions found matching cleanup criteria"
        helpers.log_msg(msg, 'ERROR')
        sys.exit(1)

    for cvid in ver_list.keys():
        # Check if there is a publish/promote already running on this content view
        locked = helpers.check_running_publish(ver_list[cvid], ver_descr[cvid])

        msg = "Cleaning content view '" + str(ver_descr[cvid]) + "'" 
        helpers.log_msg(msg, 'INFO')
        print helpers.HEADER + msg + helpers.ENDC

        # For the given content view we need to find the orphaned versions
        cvinfo = get_content_view_info(cvid)

        # Find the oldest published version
        version_list = []
        version_list_all = []
        for version in cvinfo['versions']:
            if not version['environment_ids']:
                version_list_all.append(float(version['version']))
                continue
            else:
                msg = "Found version " + str(version['version'])
                helpers.log_msg(msg, 'DEBUG')
                # Add the version id to a list
                version_list.append(float(version['version']))

        # Find the oldest 'in use' version id
        lastver = min(version_list)
        msg = "Oldest in-use version is " + str(lastver)
        helpers.log_msg(msg, 'DEBUG')

        # Find the oldest 'NOT in use' version id
        msg = "Oldest NOT-in-use version is " + str(min(version_list_all))
        helpers.log_msg(msg, 'DEBUG')

        # Find version to delete (based on keep parameter) if --ignorefirstpromoted
        version_list_all.sort()
        todelete = version_list_all[:(len(version_list_all) - int(ver_keep[cvid]))]
        msg = "Versions to remove if --ignorefirstpromoted: " + str(todelete)
        helpers.log_msg(msg, 'DEBUG')

        for version in cvinfo['versions']:
            # Find versions that are not in any environment
            if not version['environment_ids']:
                if not locked:
                    msg = "Orphan view version " + str(version['version']) + " found in '" +\
                        str(ver_descr[cvid]) + "'"
                    helpers.log_msg(msg, 'DEBUG')

                    if ignorefirstpromoted:
                        if cleanall:
                            msg = "Removing version " + str(version['version'])
                            helpers.log_msg(msg, 'INFO')
                            print helpers.HEADER + msg + helpers.ENDC
                        else:
                            if float(version['version']) in todelete:
                                # If ignorefirstpromoted delete CV
                                msg = "Removing version " + str(version['version'])
                                helpers.log_msg(msg, 'INFO')
                                print helpers.HEADER + msg + helpers.ENDC
                            else:
                                msg = "Skipping delete of version " + str(version['version']) + " due to --keep value"
                                helpers.log_msg(msg, 'INFO')
                                print msg
                                continue
                    else:
                        if float(version['version']) > float(lastver):
                            # If we have chosen to remove all orphans
                            if cleanall:
                                msg = "Removing version " + str(version['version'])
                                helpers.log_msg(msg, 'INFO')
                                print helpers.HEADER + msg + helpers.ENDC
                            else:
                                msg = "Skipping delete of version " + str(version['version'])
                                helpers.log_msg(msg, 'INFO')
                                print msg
                                continue
                        else:
                            if float(version['version']) < (lastver - float(ver_keep[cvid])):
                                msg = "Removing version " + str(version['version'])
                                helpers.log_msg(msg, 'INFO')
                                print helpers.HEADER + msg + helpers.ENDC
                            else:
                                msg = "Skipping delete of version " + str(version['version']) + " due to --keep value"
                                helpers.log_msg(msg, 'INFO')
                                print msg
                                continue

                # Delete the view version from the content view
                if not dry_run and not locked:
                    try:
                        task_id = helpers.put_json(
                            helpers.KATELLO_API + "content_views/" + str(cvid) + "/remove/",
                            json.dumps(
                                {
                                    "id": cvid,
                                    "content_view_version_ids": version['id']
                                }
                                ))['id']

                        # Wait for the task to complete
                        helpers.wait_for_task(task_id,'clean')

                        # Check if the deletion completed successfully
                        tinfo = helpers.get_task_status(task_id)
                        if tinfo['state'] != 'running' and tinfo['result'] == 'success':
                            msg = "Removal of content view version OK"
                            helpers.log_msg(msg, 'INFO')
                            print helpers.GREEN + "OK" + helpers.ENDC
                        else:
                            msg = "Failed"
                            helpers.log_msg(msg, 'ERROR')

                    except Warning:
                        msg = "Failed to initiate removal"
                        helpers.log_msg(msg, 'WARNING')

    # Exit in the case of a dry-run
    if dry_run:
        msg = "Dry run - not actually performing removal"
        helpers.log_msg(msg, 'WARNING')
        sys.exit(2)


def main(args):
    """
    Main routine
    """

    # Who is running this script?
    runuser = helpers.who_is_running()

    # Check for sane input
    parser = argparse.ArgumentParser(
        description='Cleans content views for specified organization.')
    group = parser.add_mutually_exclusive_group()
    # pylint: disable=bad-continuation
    parser.add_argument('-o', '--org', help='Organization (Uses default if not specified)',
        required=False)
    parser.add_argument('-k', '--keep', help='How many old versions to keep (only used with -a)',
        required=False)
    group.add_argument('-a', '--all', help='Clean ALL content views', required=False,
        action="store_true")
    parser.add_argument('-c', '--cleanall', help='Remove orphan versions between in-use views',
        required=False, action="store_true")
    parser.add_argument('-i', '--ignorefirstpromoted', help='Version to keep count starts from first CV, not first promoted CV',
        required=False, action="store_true")
    parser.add_argument('-d', '--dryrun', help='Dry Run - Only show what will be cleaned',
        required=False, action="store_true")

    args = parser.parse_args()

    # Log the fact we are starting
    msg = "-------- Content view cleanup started by " + runuser + " -----------"
    helpers.log_msg(msg, 'INFO')

    # Set our script variables from the input args
    if args.org:
        org_name = args.org
    else:
       org_name = helpers.ORG_NAME
    dry_run = args.dryrun
    cleanall = args.cleanall
    ignorefirstpromoted = args.ignorefirstpromoted
    if args.keep:
        keep = args.keep
    else:
        keep = "0"

    cleanup_list = []
    if not args.all:
        cleanup_list = helpers.CONFIG['cleanup']['content_views']

        if not cleanup_list:
            msg = "Cannot find cleanup configuration"
            helpers.log_msg(msg, 'ERROR')
            sys.exit(1)

        msg = "Config found for CV's " + str(cleanup_list)
        helpers.log_msg(msg, 'DEBUG')

    # Get the org_id (Validates our connection to the API)
    org_id = helpers.get_org_id(org_name)

    # Get the list of Content Views along with the latest view version in each environment
    (ver_list, ver_descr, ver_keep) = get_cv(org_id, cleanup_list, keep)

    # Clean the content views. Returns a list of task IDs.
    cleanup(ver_list, ver_descr, dry_run, runuser, ver_keep, cleanall, ignorefirstpromoted)

    # Exit cleanly
    sys.exit(0)

if __name__ == "__main__":
    try:
        main(sys.argv[1:])
    except KeyboardInterrupt, e:
        print >> sys.stderr, ("\n\nExiting on user cancel.")
        sys.exit(1)


